x/gov/migrations/v046/migrate.go

```diff
@@ -0,0 +1,97 @@
+package v046
+
+import (
+	"github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/gov/types/v1beta1"
+	"github.com/cosmos/cosmos-sdk/x/gov/types/v1beta2"
+)
+
+// ConvertToLegacyProposal takes a new proposal and attempts to convert it to the
+// legacy proposal format. This conversion is best effort. New proposal types that
+// don't have a legacy message will return a "nil" content.
+func ConvertToLegacyProposal(proposal v1beta2.Proposal) (v1beta1.Proposal, error) {
+	legacyProposal := v1beta1.Proposal{
+		ProposalId:       proposal.ProposalId,
+		Content:          nil,
+		Status:           v1beta1.ProposalStatus(proposal.Status),
+		FinalTallyResult: ConvertToLegacyTallyResult(proposal.FinalTallyResult),
+		TotalDeposit:     types.NewCoins(proposal.TotalDeposit...),
+	}
+
+	if proposal.VotingStartTime != nil {
+		legacyProposal.VotingStartTime = *proposal.VotingStartTime
+	}
+
+	if proposal.VotingEndTime != nil {
+		legacyProposal.VotingEndTime = *proposal.VotingEndTime
+	}
+
+	if proposal.SubmitTime != nil {
+		legacyProposal.SubmitTime = *proposal.SubmitTime
+	}
+
+	if proposal.DepositEndTime != nil {
+		legacyProposal.DepositEndTime = *proposal.DepositEndTime
+	}
+
+	msgs, err := proposal.GetMsgs()
+	if err != nil {
+		return v1beta1.Proposal{}, err
+	}
+	for _, msg := range msgs {
+		if legacyMsg, ok := msg.(*v1beta2.MsgExecLegacyContent); ok {
```

Comment:
if `!ok`, should we instead return an error saying "this proposal contains Msgs other than `MsgExecLegacyContent`, so cannot be converted to a legacy proposal"?

Reply:
Same if the proposal contains >=2 Msgs

Reply:
Hmm. So it seems to me that we are providing backwards compatibility but the moment someone submits a new proposal type, none of the clients will be able to read it. It's not a great form of backwards compatibility.

Reply:
Yeah. What I'm thinking is: if the user sent a `v1beta1.MsgSubmitProposal` then they can query that single proposal using `v1beta1.QueryProposal`.

In all other cases, it's going to be errors I guess

Reply:
> In all other cases, it's going to be errors I guess.

well someone can submit a `v1beta2` proposal with the legacy content only and that will be parsable by the `v1beta1` querier. But any of the new messages will return errors which I think is okay.